OpenRA.Mods.Common/MapGenerator/ActorPlan.cs

```diff
@@ -0,0 +1,149 @@
+#region Copyright & License Information
+/*
+ * Copyright (c) The OpenRA Developers and Contributors
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.MapGenerator
+{
+	/// <summary>Description of an actor to add to a map.</summary>
+	public sealed class ActorPlan
+	{
+		public readonly Map Map;
+		public readonly ActorInfo Info;
+		public readonly ActorReference Reference;
+
+		public CPos Location
+		{
+			get => Reference.Get<LocationInit>().Value;
+			set
+			{
+				Reference.RemoveAll<LocationInit>();
+				Reference.Add(new LocationInit(value));
+			}
+		}
+
+		public WPos WPosLocation
+		{
+			get => CellLayerUtils.CPosToWPos(Location, Map.Grid.Type);
+			set => Location = CellLayerUtils.WPosToCPos(value, Map.Grid.Type);
+		}
```

Comment:
this is the usual conversion we do

```suggestion
		public WPos WPosLocation
		{
			get => Map.CenterOfCell(Location);
			set => Location = Map.CellContaining(value);
		}
```

Reply:
I originally had a slight hesitation to using the Map methods. Map's CellContaining uses truncating division (rounding toward zero), whereas CellLayerUtils uses Euclidean division (flooring division, remainder always positive) to account for potentially negative WPos coordinates. For example, Map's methods will unstably map CPos(-2, -2) -> WPos(-1536, -1536) -> CPos(-1, -1) -> WPos(-512, -512) -> CPos(0, 0).

This might have been an issue in an earlier revision, but I don't think there remains any cases where this will ever happen in the current map generator logic. So, I can change this to use the map methods if preferred.

I don't know if there's any interest in changing the existing Map method behavior, but I definitely won't be changing it in this PR!